Add CLI flag & directive for clang/LLVM target triples when using Scala Native

bynux-gh opened the pull request:
This commit is meant to resolve issue #2766 by adding both a directive:
```scala
//> using nativeTargetTriple arch-mnfctr-opsys
```
and a command line option (`--native-target-triple`) that can successfully pass the target triple to Scala Native for cross-compiling projects.

As previously mentioned, this is a *very* experimental option with limited reach. Use with caution, and don't hesitate to provide suggestions for a better approach :)

bynux-gh committed:
Added CLI flag & directive for clang/LLVM target triples when using Scala Native

Finished previous commit

bynux-gh committed:
Updated reference docs

bynux-gh committed:
Used updated Scala 3 syntax per scalafmt

bynux-gh committed:
Corrected order of Scala Native directives in docs

bynux-gh committed:
Adjusted CLI option return per Scalafix recommendation

Gedochao commented on modules/cli/src/main/scala/scala/cli/commands/shared/ScalaNativeOptions.scala:
```diff
@@ -31,6 +31,11 @@ final case class ScalaNativeOptions(
 
+  @Group(HelpGroup.ScalaNative.toString)
+  @HelpMessage("Set a target triple to which Scala Native can cross-compile")
+  @Tag(tags.should)
```
```suggestion
  @Tag(tags.experimental)
```

Gedochao reviewed (changes requested):
We need integration tests, if this was to be merged. The right place would be one of:
- https://github.com/VirtusLab/scala-cli/blob/main/modules/integration/src/test/scala/scala/cli/integration/RunScalaNativeTestDefinitions.scala
- https://github.com/VirtusLab/scala-cli/blob/main/modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala

Gedochao commented:
cc @WojciechMazur

WojciechMazur commented:
The scala-native-lci integration itself LGTM, however as it was previously mentioned we need to have first a working integration test to prove that it can be actually usable. I left my concerns and mentioned problems with this flag in the comment https://github.com/VirtusLab/scala-cli/issues/2766#issuecomment-2592667517 
We should not merge it if we cannot guarantee it is somehow usable and using scala-cli we can actually create a working, cross-compiled binary and execute it.

bynux-gh committed:
Update modules/cli/src/main/scala/scala/cli/commands/shared/ScalaNativeOptions.scala

Co-authored-by: Piotr Chabelski <ged.subfan@gmail.com>

tgodzik commented:
@Gedochao do you think it would be good if we merged it under experimental? We don't actually provide any guarantees at that point.

bynux-gh committed:
Fix: System libs no longer searched w/ target triple

When --native-target-triple is invoked, the compiler will no longer pull in local libraries, which can mess with cross-compilation.

bynux-gh committed:
Set --target-triple as experimental

bynux-gh commented:
You all will be very pleased to hear that **I got cross-compilation working using this setting**! I was able to compile an AArch64 copy of a simple single-file Scala program using an x86_64 machine, then run it in qemu with minimal issues.

The last thing I hadn't accounted for is that Scala-CLI automatically adds system libraries, which is *normally* the desired behavior, but not when they're being added to a redefined sysroot. So, in the latest commit I have it overriding that behavior when a target triple is defined.

After making sure my rootfs was properly configured, I was able to point clang at it with
```sh
--native-linking "--sysroot=/usr/aarch64-linux-gnu"                                      
```

Everything else is handled automatically by `clang`/LLVM when the target triple is handed over, so in total the command I needed was:

```sh
scala-cli package Test.scala --native --native-target-triple aarch64-linux-gnu --native-linking "--sysroot=/usr/aarch64-linux-gnu" --destination $OUTDIR
```

The resulting output can be run with `qemu-aarch64` as long as it is also told where to find its libraries. It also runs without any additional input on AArch64 Arch Linux (using proot on Termux for Android on my Galaxy Note 9).

Someone who already does cross-compilation on the regular and has sysroots set up will have an easier time of making use of this for small shell programs!

Needless to say, integration tests will still be needed, and the feature should remain experimental while we only have confirmation of one particular workflow working as planned, but it's a *massive* step in the right direction. Next challenge will be using `mingw` to see if I can cross-compile Scala code to Windows :)

### TL;DR: *I have it working on two of my machines*, so we just need integration tests to demonstrate reproducible results. I was able to compile a simple one-file Scala program into a native binary that both the qemu emulator *and* a Linux installation on my Android phone can both execute without issues.

Gedochao commented:
@bynux-gh do you need help with setting up the integration test?

bynux-gh commented:
> @bynux-gh do you need help with setting up the integration test?

Hello hello, yes! I said something in the Scala Discord server but meant to bring it over to this thread. I haven't had much use for GitHub's CI system until now, so I'm not familiar with how it functions or how it integrates with test code. I could definitely use assistance setting that up and making sure the test code "tells" GitHub's systems what we want to hear.

Gedochao commented:
> Hello hello, yes! I said something in the Scala Discord server but meant to bring it over to this thread. I haven't had much use for GitHub's CI system until now, so I'm not familiar with how it functions or how it integrates with test code. I could definitely use assistance setting that up and making sure the test code "tells" GitHub's systems what we want to hear.

@bynux-gh Okay, so let's not worry about the CI too much for now. 
Let's start with adding an integration test which tests the feature and passes on your local machine, and then we'll try to figure out how to make that pass on the CI, if there are problems.

As mentioned before, the test would probably be best added to one of:
- https://github.com/VirtusLab/scala-cli/blob/main/modules/integration/src/test/scala/scala/cli/integration/RunScalaNativeTestDefinitions.scala
- https://github.com/VirtusLab/scala-cli/blob/main/modules/integration/src/test/scala/scala/cli/integration/PackageTestDefinitions.scala

More info on our integration tests can be found around here: https://github.com/VirtusLab/scala-cli/blob/main/DEV.md#run-integration-tests-with-the-jvm-launcher

As a side note, the CI is currently failing on reference docs and formatting, the commands for making those pass can be found here: https://github.com/VirtusLab/scala-cli/blob/main/CONTRIBUTING.md#rules-for-a-well-formed-pr

Let me know if you get stuck.

Gedochao commented:
@bynux-gh Is this still relevant? Do you mean to keep working on this?

Gedochao commented:
Closing this due to inactivity. Feel free to ping me to reopen if you get back to this, or just open another attempt.
